src/php/types/zval.rs

```diff
@@ -733,51 +732,122 @@ where
+                        .and_then(|val| val.try_into().ok())
+                        .or_else(|| zval.bool().map(|b| b.into()))
+                        .or_else(|| {
+                            zval.double()
+                                .map(|d| if d.is_normal() { d.floor() as _ } else { 0 })
+                        })
+                        .or_else(|| if zval.is_null() { Some(0) } else { None })
+                }
             }
-        }
+        )*
+        try_from_zval!($($t),*);
     };
 }
 
-try_from_zval!(i8, long, Long);
-try_from_zval!(i16, long, Long);
-try_from_zval!(i32, long, Long);
-try_from_zval!(i64, long, Long);
+try_from_zval!(String, f32, f64, bool);
+from_zval_long!(i8, i16, i32, i64, u8, u16, u32, u64, usize, isize);
 
-try_from_zval!(u8, long, Long);
-try_from_zval!(u16, long, Long);
-try_from_zval!(u32, long, Long);
-try_from_zval!(u64, long, Long);
+impl FromZval<'_> for String {
+    const TYPE: DataType = DataType::String;
 
-try_from_zval!(usize, long, Long);
-try_from_zval!(isize, long, Long);
+    fn from_zval(zval: &Zval) -> Option<Self> {
+        zval.string()
+    }
+
+    fn from_zval_coerce(zval: &Zval) -> Option<Self> {
+        // https://www.php.net/manual/en/language.types.string.php#language.types.string.casting
+        zval.string()
+            .or_else(|| zval.long().map(|l| l.to_string()))
+            .or_else(|| zval.double().map(|d| d.to_string()))
+            .or_else(|| zval.bool().map(|b| if b { "1" } else { "" }.to_string()))
+            .or_else(|| {
+                if zval.is_null() {
+                    Some("".to_string())
+                } else {
+                    None
+                }
+            })
+    }
+}
 
-try_from_zval!(f64, double, Double);
-try_from_zval!(bool, bool, Bool);
-try_from_zval!(String, string, String);
+impl FromZval<'_> for f64 {
+    const TYPE: DataType = DataType::Double;
+
+    fn from_zval(zval: &Zval) -> Option<Self> {
+        zval.double()
+    }
+
+    fn from_zval_coerce(zval: &Zval) -> Option<Self> {
+        zval.double()
+            .or_else(|| i64::from_zval_coerce(zval).map(|l| l as f64))
+    }
+}
 
 impl FromZval<'_> for f32 {
     const TYPE: DataType = DataType::Double;
 
     fn from_zval(zval: &Zval) -> Option<Self> {
         zval.double().map(|v| v as f32)
     }
+
+    fn from_zval_coerce(zval: &Zval) -> Option<Self> {
+        f64::from_zval_coerce(zval).map(|v| v as f32)
+    }
+}
+
+impl FromZval<'_> for bool {
+    const TYPE: DataType = DataType::Bool;
+
+    fn from_zval(zval: &Zval) -> Option<Self> {
+        zval.bool()
+    }
+
+    fn from_zval_coerce(zval: &Zval) -> Option<Self> {
+        // https://www.php.net/manual/en/language.types.boolean.php#language.types.boolean.casting
+        zval.bool()
+            .or_else(|| zval.long().map(|l| l > 0))
+            .or_else(|| zval.double().map(|d| d > 0.0))
+            .or_else(|| zval.str().map(|s| !(s.is_empty() || s == "0") || s == "1"))
```

Comment:
Any other string is considered true.
```suggestion
            .or_else(|| zval.str().map(|s| !s.is_empty() && s != "0"))
```